receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,512 @@
+		fallthrough
+	case gosnmp.Uinteger32:
+		fallthrough
+	case gosnmp.TimeTicks:
+		fallthrough
+	case gosnmp.Integer:
+		value, err := c.toInt64(pdu.Name, pdu.Value)
+		if err != nil {
+			clientSNMPData.valueType = notSupportedVal
+			clientSNMPData.value = value
+			return clientSNMPData
+		}
+
+		clientSNMPData.valueType = integerVal
+		clientSNMPData.value = value
+		return clientSNMPData
+
+	// String types
+	case gosnmp.IPAddress:
+		fallthrough
+	case gosnmp.ObjectIdentifier:
+		fallthrough
+	case gosnmp.OctetString:
+		clientSNMPData.valueType = stringVal
+		clientSNMPData.value = toString(pdu.Value)
+		return clientSNMPData
+
+	// Float types
+	case gosnmp.OpaqueFloat:
+		fallthrough
+	case gosnmp.OpaqueDouble:
+		value, err := c.toFloat64(pdu.Name, pdu.Value)
+		if err != nil {
+			clientSNMPData.valueType = notSupportedVal
+			clientSNMPData.value = value
+			return clientSNMPData
+		}
+
+		clientSNMPData.valueType = floatVal
+		clientSNMPData.value = value
+		return clientSNMPData
+
+	// Not supported types either because gosnmp doesn't support them
+	// or they are a type that doesn't translate well to OTEL
+	case gosnmp.UnknownType:
+		fallthrough
+	case gosnmp.Counter64:
+		fallthrough
+	case gosnmp.NsapAddress:
+		fallthrough
+	case gosnmp.ObjectDescription:
+		fallthrough
+	case gosnmp.BitString:
+		fallthrough
+	case gosnmp.NoSuchObject:
+		fallthrough
+	case gosnmp.NoSuchInstance:
+		fallthrough
+	case gosnmp.EndOfMibView:
+		fallthrough
+	case gosnmp.Opaque:
+		fallthrough
+	case gosnmp.Null:
+		fallthrough
+	case gosnmp.Boolean:
+		fallthrough
+	default:
+		clientSNMPData.valueType = notSupportedVal
+		clientSNMPData.value = pdu.Value
+		return clientSNMPData
+	}
+}
+
+// toInt64 converts SnmpPDU.Value to int64, or returns an error for
+// non int-like types or a uint64.
+//
+// This is a convenience function to make working with SnmpPDU's easier - it
+// reduces the need for type assertions. A int64 is convenient, as SNMP can
+// return int32, uint32, and int64.
+func (c snmpClient) toInt64(name string, value interface{}) (int64, error) {
+	var val int64
+
+	switch value := value.(type) { // shadow
+	case int:
+		val = int64(value)
+	case int8:
+		val = int64(value)
+	case int16:
+		val = int64(value)
+	case int32:
+		val = int64(value)
+	case int64:
+		val = value
+	case uint:
+		val = int64(value)
+	case uint8:
+		val = int64(value)
+	case uint16:
+		val = int64(value)
+	case uint32:
+		val = int64(value)
+	default:
+		return 0, fmt.Errorf("incompatible type while converting OID '%s' data to int64", name)
+	}
+
+	return val, nil
+}
+
+// toFloat64 converts SnmpPDU.Value to float64, or returns an error for non
+// float like types.
+//
+// This is a convenience function to make working with SnmpPDU's easier - it
+// reduces the need for type assertions. A float64 is convenient, as SNMP can
+// return float32 and float64.
+func (c snmpClient) toFloat64(name string, value interface{}) (float64, error) {
+	var val float64
```

Comment:
I think this allocation is unnecessary. Just return directly from cases.

Reply:
@djaglowski sure i don't mind changing these